Roomモデルのメソッドに関するテストを作成する

uchihiro04 opened the pull request:
## 概要 
Roomモデルのメソッドに関するテストを作成しました。
テストを作成したメソッドは以下の通りです。
* #average_score
* #average_total_score
* .score_entered?
* #reviewer_names

## スクリーンショット
### テストを実行
<img width="625" alt="スクリーンショット 2023-07-28 17 44 32" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/54be1208-afe7-4df7-b506-a93f13839192">

## 関連Issue
- #145 

Close #145

uchihiro04 committed:
#145 average_scoreメソッドのテストを作成

Roomレコードにスコアが紐づいているかどうかで条件分岐させています。

uchihiro04 committed:
#145 to_paramメソッドに関するコメント追加

HouseViewingモデルのインスタンスメソッドであるto_paramメソッドについて、コメントを追加しました。

Yuma-Nagaoka commented on spec/models/room_spec.rb:
```diff
@@ -0,0 +1,25 @@
+
+    context 'スコアが存在する場合' do
+      it 'スコアの平均値を返すこと' do
+        room.scores = scores
```
この段階でroomにscoresが紐づいていないのは、Roomレコードが作られてインスタンスがroomに格納された後にscoresがつくられているため、room変数にはscoresとの紐付きの情報がないからです。
speeeだとこういう場合はroom.reloadを使って読み込み直しています。
`expect(room.reload.average_score(Score::EVALUATION_ITEMS.first)).to eq 2.0`
といった具合になります。

uchihiro04 replied:
いただいたコメントおよびハドルにより、以下を理解することができました。ありがとうございます。
* `room.reload`でroomに関する情報を更新できる
* `let`は定義後に改めて使用しないと実行されない（すぐに実行する場合は`let!`を使う）

reloadを用いた形にコードを修正しました。該当するコミットは[62a6940](https://github.com/uchihiro04/house-viewing-score-log/pull/148/commits/62a694056ff17ed2be0f42a2576c5f71d7ef1ff1)です。

Yuma-Nagaoka commented on spec/models/room_spec.rb:
```diff
@@ -0,0 +1,25 @@
+
+    context 'スコアが存在しない場合' do
+      it '0を返すこと' do
+        expect(room.average_score(Score::EVALUATION_ITEMS.first)).to eq 0
```
ここもreloadしちゃってください

uchihiro04 replied:
いただいたレビューをもとにreloadを用いた形にコードを修正しました。該当するコミットは[d29e405](https://github.com/uchihiro04/house-viewing-score-log/pull/148/commits/d29e405bbaa56e06e0dcb4e5deb0bb3276f66c8d)です。

Yuma-Nagaoka commented on spec/models/room_spec.rb:
```diff
@@ -0,0 +1,25 @@
+  describe '#average_score' do
+    let(:house_viewing) { create(:house_viewing) }
+    let(:room) { create(:room, house_viewing:) }
+    let(:scores) { create_list('score', 3, room:) }
```
作成するスコアごとにカラムの値を変えたい場合は以下のように書くのがいいかなと思います

```
before do
   test_values.each do |test_value| 
      create(:score, room:, living_room: test_value)
   end
end

context 'スコアが存在する場合' do
   context 'テストケース1' do
      let(:test_values) { [1, 2, 3] }
      it 'スコアの平均値を返すこと' do
      ...
      end
   end

   context 'テストケース2' do
    ...
   end
end

context 'スコアが存在しない場合' do
   let(:test_values) { [] }
   it '0を返すこと' do
        expect(room.reload.average_score(:living_room)).to eq 0
   end
end
```

uchihiro04 replied:
記載いただいたコードを活用することで、`spec/factories/score.rb`でsequenceを使う必要がなくなりました。また、テスト時に使う評価項目のスコアも柔軟性が出そうです。ありがとうございます。

いただいたレビューをもとに、コードを修正しました。該当するコミットは[88920b3](https://github.com/uchihiro04/house-viewing-score-log/pull/148/commits/88920b36df702fa1728e45e07102875894e5d03d)です。

Yuma-Nagaoka commented on spec/factories/score.rb:
```diff
@@ -0,0 +1,15 @@
+
+FactoryBot.define do
+  factory :score do
+    sequence(:reviewer_name) { |n| "ユーザー#{n}" }
```
sequenceを使ったことがなかったのですが、factoryが呼び出されるたびに1から増えていくようなものですね。
すみません、create_listで指定してしまったのでこういう実装になったんだと思うのですが、factoryで作成するレコードはランダムせいがあった方が好ましいので、スコアカラムに対しては{ rand(5) }のような形がいいと思います。

### tips
以下正規表現でサクッと置換できると思います
正規表現 `sequence\(:(.*)\).*`
置換の指定 `$1 { rand(5) }`

Yuma-Nagaoka replied:
markdown認識されている部分があったので再掲します
正規表現 `sequence\(:(.*)\).*`
置換の指定 `$1 { rand(5) }`

uchihiro04 replied:
確かにランダムにしたほうが、ダミーデータとして使う場合に適していそうです。

いただいたレビューをもとに、Scoreの評価項目についてsequenceからrandメソッドを使う形に変更しました。該当するコミットは[3cea843](https://github.com/uchihiro04/house-viewing-score-log/pull/148/commits/3cea8431401fbec41908335e33f9ed7f92db3d73)です。

uchihiro04 committed:
#145 スコアの平均値を返すテストについて、reloadを使用する形に変更

#average_scoreメソッドのスコアの平均値を返すテストについて、reloadを用いてroomの情報を更新する形に変更しました。

uchihiro04 committed:
#145 スコアが存在しない場合のテストについて、reloadを使用する形に変更

#average_scoreメソッドの、スコアが存在しない場合のテストについて、reloadを用いてroomの情報を更新する形に変更しました。

uchihiro04 committed:
#145 Scoreのテストデータについて、評価項目の得点を修正

Scoreのテストデータの評価項目について、sequenceを使う方法から、randメソッドを使う方法に修正しました。

uchihiro04 committed:
#145 Scoreデータの作成方法を変更

Scoreデータについて、create_listを用いた方法から、beforeとcreateを用いた方法に変更しました。

Yuma-Nagaoka commented on spec/factories/house_viewing.rb:
```diff
@@ -0,0 +1,7 @@
+
+FactoryBot.define do
+  factory :house_viewing do
+    uuid { SecureRandom.uuid }
```
コールバックで作られるのでuuidの処理はいらないと思います。

uchihiro04 replied:
確かにコールバックでuuidは作成されるため、ここで処理する必要はないですね……。`SecureRandom.uuid`は削除したいと思います。

rubocopの`Lint/EmptyBlock: Empty block detected.`という指摘を回避するために、`uuid { SecureRandom.uuid }`を`uuid { nil }`に変更しました。該当するコミットは[f9c49f8](https://github.com/uchihiro04/house-viewing-score-log/pull/148/commits/f9c49f8b1bd1110c9fabc6199fb3abd1af96452d)です。

Yuma-Nagaoka commented on spec/models/room_spec.rb:
```diff
@@ -0,0 +1,33 @@
+
+    before do
+      test_values.each do |test_value|
+        create(:score, room:, living_room: test_value)
```
書き方分からなかったので`living_room: test_value`で提案したんですけど、
`Score::EVALUATION_ITEMS.first => test_value`と書けば引数のキーを変数で渡せるみたいなので
`room.reload.average_score(Score::EVALUATION_ITEMS.first)`と統一させるためこっちでお願いします！

uchihiro04 replied:
試してみたところ、`"#{Score::EVALUATION_ITEMS.first}": test_value`でも渡せることが分かりましたがご提案いただいたコードの方がスッキリしていますね……。

いただいたレビューをもとに、createの第3引数を`Score::EVALUATION_ITEMS.first => test_value`に変更しました。該当するコミットは[6728b12](https://github.com/uchihiro04/house-viewing-score-log/pull/148/commits/6728b12e6e27cd7c1054e91a342b321d9cf49b6b)です。

Yuma-Nagaoka commented on spec/factories/rooms.rb:
```diff
@@ -0,0 +1,7 @@
+
+FactoryBot.define do
+  factory :room do
+    name { '1件目' }
```
fakerというgemを用いて名前もある程度ランダムにできるんですが、そんなに難しくないし案件で触れば理解できるので、一旦このままでいいと思います。

uchihiro04 replied:
ありがとうございます。
fakerはフィヨルドで1回使ったことがありました。ダミーの種類が多かった記憶があります。

今回はご提案いただいた通り、fakerを使わずこのまま進めていきたいと思います。

uchihiro04 committed:
#145 createメソッドの第3引数を変更

createメソッドの第3引数について、キーに変数を用いる形に変更しました。

uchihiro04 committed:
#145 Scoreデータ作成時の不要な処理の削除

コールバック時にuuidを生成するため、spec/factories/house_viewing.rb内でのuuidの作成に関するコードを削除しました。

uchihiro04 committed:
#145 average_total_scoreメソッドについて、スコアが存在する場合のテストを作成

uchihiro04 committed:
#145 average_total_scoreメソッドについて、スコアが存在しない場合のテストを作成

uchihiro04 committed:
#145 .score_entered?に関するテストの作成

uchihiro04 committed:
#145 rubocopの設定変更

rubocopのMetrics/BlockLengthについて、RSpecではブロック内の行数が多くなるためrubocopのチェック項目から削除しました。

uchihiro04 committed:
#145 ファイル名の変更

rails g rspec:modelを実行した際に作成されるファイルの命名規則と同じ形にするために、spec/factories/score.rbからspec/factories/scores.rbにファイル名を変更しました。

uchihiro04 committed:
#145 beforeのスコープを変更

before内のコードが不要なテストが存在するため、必要なdescribeにだけbefore内のコードが適用される形に変更しました。

uchihiro04 committed:
#145 reviewer_namesメソッドに関するテストを作成

Yuma-Nagaoka commented on spec/models/room_spec.rb:
```diff
@@ -0,0 +1,100 @@
+  describe '#average_score' do
+    before do
+      test_values.each do |test_value|
+        evaluation_items = { room: }
```
この書き方初めて見ました:eyes: 
rails7からの書き方ですかね

uchihiro04 replied:
rubocopで指摘を受けたため`evaluation_items = { room: }`という書き方に修正をした形です。自分も今まで知りませんでした……。

調べてみたところ、Ruby3.1でハッシュリテラルを省略する記法が導入されていて、rubocopのチェックもそれに準じているとのことでした。

参考リンク
[Ruby 3\.1: ES6風のハッシュリテラル省略記法が追加（翻訳）｜TechRacho by BPS株式会社](https://techracho.bpsinc.jp/hachi8833/2021_11_24/113383)
[RuboCop 1\.24\.0 と Ruby 3\.1 対応 \- koicの日記](https://koic.hatenablog.com/entry/rubocop-supports-ruby-3-1-syntax)
[Support Ruby 3\.1's hash value omission for \`Style/HashSyntax\` by koic · Pull Request \#10289 · rubocop/rubocop](https://github.com/rubocop/rubocop/pull/10289)
